// @hyperlane-xyz/sdk/dist/HyperlaneApp.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HyperlaneApp = void 0;
const debug_1 = __importDefault(require("debug"));
const contracts_1 = require("./contracts");
const MultiGeneric_1 = require("./utils/MultiGeneric");
const objects_1 = require("./utils/objects");
class HyperlaneApp extends MultiGeneric_1.MultiGeneric {
    constructor(contractsMap, multiProvider, logger = (0, debug_1.default)('hyperlane:App')) {
        const connectedContractsMap = (0, objects_1.objMap)(contractsMap, (chain, contracts) => (0, contracts_1.connectContracts)(contracts, multiProvider.getSignerOrProvider(chain)));
        super(connectedContractsMap);
        this.contractsMap = contractsMap;
        this.multiProvider = multiProvider;
        this.logger = logger;
    }
    getContracts(chain) {
        return this.get(chain);
    }
    getAddresses(chain) {
        return (0, contracts_1.serializeContracts)(this.get(chain));
    }
}
exports.HyperlaneApp = HyperlaneApp;
