// @hyperlane-xyz/utils/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.trace = exports.debug = exports.log = exports.warn = exports.error = exports.logging = exports.BaseValidator = exports.Validator = exports.utils = exports.types = void 0;
exports.types = __importStar(require("./src/types"));
exports.utils = __importStar(require("./src/utils"));
var validator_1 = require("./src/validator");
Object.defineProperty(exports, "Validator", { enumerable: true, get: function () { return validator_1.Validator; } });
Object.defineProperty(exports, "BaseValidator", { enumerable: true, get: function () { return validator_1.BaseValidator; } });
exports.logging = __importStar(require("./src/logging"));
var logging_1 = require("./src/logging");
Object.defineProperty(exports, "error", { enumerable: true, get: function () { return logging_1.error; } });
Object.defineProperty(exports, "warn", { enumerable: true, get: function () { return logging_1.warn; } });
Object.defineProperty(exports, "log", { enumerable: true, get: function () { return logging_1.log; } });
Object.defineProperty(exports, "debug", { enumerable: true, get: function () { return logging_1.debug; } });
Object.defineProperty(exports, "trace", { enumerable: true, get: function () { return logging_1.trace; } });
