// @hyperlane-xyz/core/dist/factories/contracts/interfaces/ILiquidityLayerMessageRecipient__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ILiquidityLayerMessageRecipient__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint32",
                name: "_origin",
                type: "uint32",
            },
            {
                internalType: "bytes32",
                name: "_sender",
                type: "bytes32",
            },
            {
                internalType: "bytes",
                name: "_message",
                type: "bytes",
            },
            {
                internalType: "address",
                name: "_token",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "_amount",
                type: "uint256",
            },
        ],
        name: "handleWithTokens",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class ILiquidityLayerMessageRecipient__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ILiquidityLayerMessageRecipient__factory = ILiquidityLayerMessageRecipient__factory;
ILiquidityLayerMessageRecipient__factory.abi = _abi;
