// @hyperlane-xyz/core/dist/factories/contracts/interfaces/middleware/IInterchainQueryRouter__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IInterchainQueryRouter__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint32",
                name: "_destination",
                type: "uint32",
            },
            {
                internalType: "address",
                name: "_to",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "_data",
                type: "bytes",
            },
            {
                internalType: "bytes",
                name: "_callback",
                type: "bytes",
            },
        ],
        name: "query",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint32",
                name: "_destination",
                type: "uint32",
            },
            {
                components: [
                    {
                        components: [
                            {
                                internalType: "bytes32",
                                name: "to",
                                type: "bytes32",
                            },
                            {
                                internalType: "bytes",
                                name: "data",
                                type: "bytes",
                            },
                        ],
                        internalType: "struct CallLib.StaticCall",
                        name: "_call",
                        type: "tuple",
                    },
                    {
                        internalType: "bytes",
                        name: "callback",
                        type: "bytes",
                    },
                ],
                internalType: "struct CallLib.StaticCallWithCallback[]",
                name: "calls",
                type: "tuple[]",
            },
        ],
        name: "query",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IInterchainQueryRouter__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IInterchainQueryRouter__factory = IInterchainQueryRouter__factory;
IInterchainQueryRouter__factory.abi = _abi;
