// @hyperlane-xyz/sdk/dist/consts/chainMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValidChainMetadata = exports.testnetChainsMetadata = exports.mainnetChainsMetadata = exports.chainIdToMetadata = exports.wagmiChainMetadata = exports.chainMetadata = exports.test3 = exports.test2 = exports.test1 = exports.gnosis = exports.polygon = exports.optimismgoerli = exports.optimism = exports.mumbai = exports.moonbeam = exports.moonbasealpha = exports.sepolia = exports.goerli = exports.fuji = exports.ethereum = exports.celo = exports.bsctestnet = exports.bsc = exports.avalanche = exports.arbitrumgoerli = exports.arbitrum = exports.alfajores = exports.xDaiToken = exports.maticToken = exports.etherToken = exports.celoToken = exports.bnbToken = exports.avaxToken = exports.ChainMetadataSchema = exports.ProtocolType = exports.ExplorerFamily = void 0;
const zod_1 = require("zod");
const objects_1 = require("../utils/objects");
const wagmi_1 = require("../utils/wagmi");
const chains_1 = require("./chains");
var ExplorerFamily;
(function (ExplorerFamily) {
    ExplorerFamily["Etherscan"] = "etherscan";
    ExplorerFamily["Blockscout"] = "blockscout";
    ExplorerFamily["Other"] = "other";
})(ExplorerFamily = exports.ExplorerFamily || (exports.ExplorerFamily = {}));
var ProtocolType;
(function (ProtocolType) {
    ProtocolType["Ethereum"] = "ethereum";
    ProtocolType["Sealevel"] = "sealevel";
    ProtocolType["Fuel"] = "fuel";
})(ProtocolType = exports.ProtocolType || (exports.ProtocolType = {}));
/**
 * Zod schema for ChainMetadata validation
 * Keep in sync with ChainMetadata above
 */
exports.ChainMetadataSchema = zod_1.z.object({
    chainId: zod_1.z.number().positive(),
    domainId: zod_1.z.number().positive().optional(),
    name: zod_1.z.string(),
    protocol: zod_1.z.string(),
    displayName: zod_1.z.string().optional(),
    displayNameShort: zod_1.z.string().optional(),
    nativeToken: zod_1.z
        .object({
        name: zod_1.z.string(),
        symbol: zod_1.z.string(),
        decimals: zod_1.z.number().positive(),
    })
        .optional(),
    publicRpcUrls: zod_1.z
        .array(zod_1.z.object({
        http: zod_1.z.string().url(),
        webSocket: zod_1.z.string().optional(),
        pagination: zod_1.z
            .object({
            maxBlockRange: zod_1.z.number().positive().optional(),
            minBlockNumber: zod_1.z.number().positive().optional(),
            maxBlockAge: zod_1.z.number().positive().optional(),
        })
            .optional(),
        retry: zod_1.z
            .object({
            maxRequests: zod_1.z.number().positive(),
            baseRetryMs: zod_1.z.number().positive(),
        })
            .optional(),
    }))
        .nonempty(),
    blockExplorers: zod_1.z
        .array(zod_1.z.object({
        name: zod_1.z.string(),
        url: zod_1.z.string().url(),
        apiUrl: zod_1.z.string().url(),
        apiKey: zod_1.z.string().optional(),
        family: zod_1.z.nativeEnum(ExplorerFamily).optional(),
    }))
        .optional(),
    blocks: zod_1.z
        .object({
        confirmations: zod_1.z.number(),
        reorgPeriod: zod_1.z.number().optional(),
        estimateBlockTime: zod_1.z.number().positive().optional(),
    })
        .optional(),
    transactionOverrides: zod_1.z.object({}).optional(),
    gasCurrencyCoinGeckoId: zod_1.z.string().optional(),
    gnosisSafeTransactionServiceUrl: zod_1.z.string().optional(),
    isTestnet: zod_1.z.boolean().optional(),
});
/**
 * Common native currencies
 */
exports.avaxToken = {
    decimals: 18,
    name: 'Avalanche',
    symbol: 'AVAX',
};
exports.bnbToken = {
    decimals: 18,
    name: 'BNB',
    symbol: 'BNB',
};
exports.celoToken = {
    decimals: 18,
    name: 'CELO',
    symbol: 'CELO',
};
exports.etherToken = { name: 'Ether', symbol: 'ETH', decimals: 18 };
exports.maticToken = { name: 'MATIC', symbol: 'MATIC', decimals: 18 };
exports.xDaiToken = { name: 'xDai', symbol: 'xDai', decimals: 18 };
/**
 * Chain metadata
 */
exports.alfajores = {
    chainId: 44787,
    name: chains_1.Chains.alfajores,
    protocol: ProtocolType.Ethereum,
    displayName: 'Alfajores',
    nativeToken: exports.celoToken,
    publicRpcUrls: [{ http: 'https://alfajores-forno.celo-testnet.org' }],
    blockExplorers: [
        {
            name: 'CeloScan',
            url: 'https://alfajores.celoscan.io',
            apiUrl: 'https://api-alfajores.celoscan.io/api',
            family: ExplorerFamily.Etherscan,
        },
        {
            name: 'Blockscout',
            url: 'https://explorer.celo.org/alfajores',
            apiUrl: 'https://explorer.celo.org/alfajores/api',
            family: ExplorerFamily.Blockscout,
        },
    ],
    blocks: {
        confirmations: 1,
        reorgPeriod: 0,
        estimateBlockTime: 5,
    },
    isTestnet: true,
};
exports.arbitrum = {
    chainId: 42161,
    name: chains_1.Chains.arbitrum,
    protocol: ProtocolType.Ethereum,
    displayName: 'Arbitrum',
    nativeToken: exports.etherToken,
    publicRpcUrls: [{ http: 'https://arb1.arbitrum.io/rpc' }],
    blockExplorers: [
        {
            name: 'Arbiscan',
            url: 'https://arbiscan.io',
            apiUrl: 'https://api.arbiscan.io/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 1,
        reorgPeriod: 0,
        estimateBlockTime: 3,
    },
    gasCurrencyCoinGeckoId: 'ethereum',
    gnosisSafeTransactionServiceUrl: 'https://safe-transaction.arbitrum.gnosis.io/',
};
exports.arbitrumgoerli = {
    chainId: 421613,
    name: chains_1.Chains.arbitrumgoerli,
    protocol: ProtocolType.Ethereum,
    displayName: 'Arbitrum Goerli',
    displayNameShort: 'Arb. Goerli',
    nativeToken: exports.etherToken,
    publicRpcUrls: [{ http: 'https://goerli-rollup.arbitrum.io/rpc' }],
    blockExplorers: [
        {
            name: 'Arbiscan',
            url: 'https://goerli.arbiscan.io',
            apiUrl: 'https://api-goerli.arbiscan.io/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 1,
        reorgPeriod: 1,
        estimateBlockTime: 3,
    },
    isTestnet: true,
};
exports.avalanche = {
    chainId: 43114,
    name: chains_1.Chains.avalanche,
    protocol: ProtocolType.Ethereum,
    displayName: 'Avalanche',
    nativeToken: exports.avaxToken,
    publicRpcUrls: [
        {
            http: 'https://api.avax.network/ext/bc/C/rpc',
            pagination: {
                maxBlockRange: 100000,
                minBlockNumber: 6765067,
            },
        },
    ],
    blockExplorers: [
        {
            name: 'SnowTrace',
            url: 'https://snowtrace.io',
            apiUrl: 'https://api.snowtrace.io/api',
            family: ExplorerFamily.Other,
        },
    ],
    blocks: {
        confirmations: 3,
        reorgPeriod: 3,
        estimateBlockTime: 2,
    },
    gasCurrencyCoinGeckoId: 'avalanche-2',
    gnosisSafeTransactionServiceUrl: 'https://safe-transaction.avalanche.gnosis.io/',
};
exports.bsc = {
    chainId: 56,
    name: chains_1.Chains.bsc,
    protocol: ProtocolType.Ethereum,
    displayName: 'Binance Smart Chain',
    displayNameShort: 'Binance',
    nativeToken: exports.bnbToken,
    publicRpcUrls: [
        { http: 'https://bsc-dataseed.binance.org' },
        { http: 'https://rpc.ankr.com/bsc' },
    ],
    blockExplorers: [
        {
            name: 'BscScan',
            url: 'https://bscscan.com',
            apiUrl: 'https://api.bscscan.com/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 1,
        reorgPeriod: 15,
        estimateBlockTime: 3,
    },
    gasCurrencyCoinGeckoId: 'binancecoin',
    gnosisSafeTransactionServiceUrl: 'https://safe-transaction.bsc.gnosis.io/',
};
exports.bsctestnet = {
    chainId: 97,
    name: chains_1.Chains.bsctestnet,
    protocol: ProtocolType.Ethereum,
    displayName: 'BSC Testnet',
    nativeToken: exports.bnbToken,
    publicRpcUrls: [{ http: 'https://data-seed-prebsc-1-s3.binance.org:8545' }],
    blockExplorers: [
        {
            name: 'BscScan',
            url: 'https://testnet.bscscan.com',
            apiUrl: 'https://api-testnet.bscscan.com/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 1,
        reorgPeriod: 9,
        estimateBlockTime: 3,
    },
    isTestnet: true,
};
exports.celo = {
    chainId: 42220,
    name: chains_1.Chains.celo,
    protocol: ProtocolType.Ethereum,
    displayName: 'Celo',
    nativeToken: exports.celoToken,
    publicRpcUrls: [{ http: 'https://forno.celo.org' }],
    blockExplorers: [
        {
            name: 'CeloScan',
            url: 'https://celoscan.io',
            apiUrl: 'https://api.celoscan.io/api',
            family: ExplorerFamily.Etherscan,
        },
        {
            name: 'Blockscout',
            url: 'https://explorer.celo.org',
            apiUrl: 'https://explorer.celo.org/mainnet/api',
            family: ExplorerFamily.Blockscout,
        },
    ],
    blocks: {
        confirmations: 1,
        reorgPeriod: 0,
        estimateBlockTime: 5,
    },
    gnosisSafeTransactionServiceUrl: 'https://transaction-service.gnosis-safe-staging.celo-networks-dev.org',
};
exports.ethereum = {
    chainId: 1,
    name: chains_1.Chains.ethereum,
    protocol: ProtocolType.Ethereum,
    displayName: 'Ethereum',
    nativeToken: exports.etherToken,
    publicRpcUrls: [
        { http: 'https://mainnet.infura.io/v3/9aa3d95b3bc440fa88ea12eaa4456161' },
        { http: 'https://cloudflare-eth.com' },
    ],
    blockExplorers: [
        {
            name: 'Etherscan',
            url: 'https://etherscan.io',
            apiUrl: 'https://api.etherscan.io/api',
            family: ExplorerFamily.Etherscan,
        },
        {
            name: 'Blockscout',
            url: 'https://blockscout.com/eth/mainnet',
            apiUrl: 'https://blockscout.com/eth/mainnet/api',
            family: ExplorerFamily.Blockscout,
        },
    ],
    blocks: {
        confirmations: 7,
        reorgPeriod: 14,
        estimateBlockTime: 13,
    },
    gnosisSafeTransactionServiceUrl: 'https://safe-transaction.gnosis.io',
};
exports.fuji = {
    chainId: 43113,
    name: chains_1.Chains.fuji,
    protocol: ProtocolType.Ethereum,
    displayName: 'Fuji',
    nativeToken: exports.avaxToken,
    publicRpcUrls: [
        {
            http: 'https://api.avax-test.network/ext/bc/C/rpc',
            pagination: { maxBlockRange: 2048 },
        },
    ],
    blockExplorers: [
        {
            name: 'SnowTrace',
            url: 'https://testnet.snowtrace.io',
            apiUrl: 'https://api-testnet.snowtrace.io/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 3,
        reorgPeriod: 3,
        estimateBlockTime: 2,
    },
    isTestnet: true,
};
exports.goerli = {
    chainId: 5,
    name: chains_1.Chains.goerli,
    protocol: ProtocolType.Ethereum,
    displayName: 'Goerli',
    nativeToken: exports.etherToken,
    publicRpcUrls: [
        { http: 'https://goerli.infura.io/v3/9aa3d95b3bc440fa88ea12eaa4456161' },
        { http: 'https://rpc.ankr.com/eth_goerli' },
        { http: 'https://eth-goerli.public.blastapi.io' },
    ],
    blockExplorers: [
        {
            name: 'Etherscan',
            url: 'https://goerli.etherscan.io',
            apiUrl: 'https://api-goerli.etherscan.io/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 1,
        reorgPeriod: 2,
        estimateBlockTime: 13,
    },
    isTestnet: true,
};
exports.sepolia = {
    chainId: 11155111,
    name: chains_1.Chains.sepolia,
    protocol: ProtocolType.Ethereum,
    displayName: 'Sepolia',
    nativeToken: exports.etherToken,
    publicRpcUrls: [
        { http: 'https://endpoints.omniatech.io/v1/eth/sepolia/public' },
        { http: 'https://rpc.sepolia.org' },
    ],
    blockExplorers: [
        {
            name: 'Etherscan',
            url: 'https://sepolia.etherscan.io',
            apiUrl: 'https://api-sepolia.etherscan.io/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 1,
        reorgPeriod: 2,
        estimateBlockTime: 13,
    },
    isTestnet: true,
};
exports.moonbasealpha = {
    chainId: 1287,
    name: chains_1.Chains.moonbasealpha,
    protocol: ProtocolType.Ethereum,
    displayName: 'Moonbase Alpha',
    displayNameShort: 'Moonbase',
    nativeToken: {
        decimals: 18,
        name: 'DEV',
        symbol: 'DEV',
    },
    publicRpcUrls: [{ http: 'https://rpc.api.moonbase.moonbeam.network' }],
    blockExplorers: [
        {
            name: 'MoonScan',
            url: 'https://moonbase.moonscan.io',
            apiUrl: 'https://api-moonbase.moonscan.io/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 1,
        reorgPeriod: 1,
        estimateBlockTime: 12,
    },
    isTestnet: true,
};
exports.moonbeam = {
    chainId: 1284,
    name: chains_1.Chains.moonbeam,
    protocol: ProtocolType.Ethereum,
    displayName: 'Moonbeam',
    nativeToken: {
        decimals: 18,
        name: 'GLMR',
        symbol: 'GLMR',
    },
    publicRpcUrls: [{ http: 'https://rpc.api.moonbeam.network' }],
    blockExplorers: [
        {
            name: 'MoonScan',
            url: 'https://moonscan.io',
            apiUrl: 'https://api-moonbeam.moonscan.io/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 2,
        reorgPeriod: 2,
        estimateBlockTime: 12,
    },
    gnosisSafeTransactionServiceUrl: 'https://transaction.multisig.moonbeam.network',
};
exports.mumbai = {
    chainId: 80001,
    name: chains_1.Chains.mumbai,
    protocol: ProtocolType.Ethereum,
    displayName: 'Mumbai',
    nativeToken: exports.maticToken,
    publicRpcUrls: [
        {
            http: 'https://rpc.ankr.com/polygon_mumbai',
            pagination: {
                // eth_getLogs and eth_newFilter are limited to a 10,000 blocks range
                maxBlockRange: 10000,
                minBlockNumber: 22900000,
            },
        },
        {
            http: 'https://matic-mumbai.chainstacklabs.com',
        },
    ],
    blockExplorers: [
        {
            name: 'PolygonScan',
            url: 'https://mumbai.polygonscan.com',
            apiUrl: 'https://api-testnet.polygonscan.com/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 3,
        reorgPeriod: 32,
        estimateBlockTime: 5,
    },
    isTestnet: true,
};
exports.optimism = {
    chainId: 10,
    name: chains_1.Chains.optimism,
    protocol: ProtocolType.Ethereum,
    displayName: 'Optimism',
    nativeToken: exports.etherToken,
    publicRpcUrls: [{ http: 'https://mainnet.optimism.io' }],
    blockExplorers: [
        {
            name: 'Etherscan',
            url: 'https://optimistic.etherscan.io',
            apiUrl: 'https://api-optimistic.etherscan.io/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 1,
        reorgPeriod: 0,
        estimateBlockTime: 3,
    },
    gasCurrencyCoinGeckoId: 'ethereum',
    gnosisSafeTransactionServiceUrl: 'https://safe-transaction.optimism.gnosis.io/',
};
exports.optimismgoerli = {
    chainId: 420,
    name: chains_1.Chains.optimismgoerli,
    protocol: ProtocolType.Ethereum,
    displayName: 'Optimism Goerli',
    displayNameShort: 'Opt. Goerli',
    nativeToken: exports.etherToken,
    publicRpcUrls: [{ http: 'https://goerli.optimism.io' }],
    blockExplorers: [
        {
            name: 'Etherscan',
            url: 'https://goerli-optimism.etherscan.io',
            apiUrl: 'https://api-goerli-optimism.etherscan.io/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 1,
        reorgPeriod: 1,
        estimateBlockTime: 3,
    },
    isTestnet: true,
};
exports.polygon = {
    chainId: 137,
    name: chains_1.Chains.polygon,
    protocol: ProtocolType.Ethereum,
    displayName: 'Polygon',
    nativeToken: exports.etherToken,
    publicRpcUrls: [
        {
            http: 'https://rpc-mainnet.matic.quiknode.pro',
            pagination: {
                // Needs to be low to avoid RPC timeouts
                maxBlockRange: 10000,
                minBlockNumber: 19657100,
            },
        },
        { http: 'https://polygon-rpc.com' },
    ],
    blockExplorers: [
        {
            name: 'PolygonScan',
            url: 'https://polygonscan.com',
            apiUrl: 'https://api.polygonscan.com/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 200,
        reorgPeriod: 256,
        estimateBlockTime: 2,
    },
    gasCurrencyCoinGeckoId: 'matic-network',
    gnosisSafeTransactionServiceUrl: 'https://safe-transaction.polygon.gnosis.io/',
};
exports.gnosis = {
    chainId: 100,
    name: chains_1.Chains.gnosis,
    protocol: ProtocolType.Ethereum,
    displayName: 'Gnosis',
    nativeToken: exports.xDaiToken,
    publicRpcUrls: [
        {
            http: 'https://rpc.gnosischain.com',
            pagination: {
                maxBlockRange: 10000,
                minBlockNumber: 25997478,
            },
        },
    ],
    blockExplorers: [
        {
            name: 'GnosisScan',
            url: 'https://gnosisscan.io',
            apiUrl: 'https://api.gnosisscan.io/api',
            family: ExplorerFamily.Etherscan,
        },
    ],
    blocks: {
        confirmations: 1,
        reorgPeriod: 14,
        estimateBlockTime: 5,
    },
    gasCurrencyCoinGeckoId: 'xdai',
    gnosisSafeTransactionServiceUrl: 'https://safe-transaction.xdai.gnosis.io/',
};
exports.test1 = {
    chainId: 13371,
    name: chains_1.Chains.test1,
    protocol: ProtocolType.Ethereum,
    displayName: 'Test 1',
    nativeToken: exports.etherToken,
    publicRpcUrls: [{ http: 'http://localhost:8545' }],
    blockExplorers: [],
    blocks: {
        confirmations: 1,
        reorgPeriod: 0,
        estimateBlockTime: 3,
    },
    isTestnet: true,
};
exports.test2 = {
    chainId: 13372,
    name: chains_1.Chains.test2,
    protocol: ProtocolType.Ethereum,
    displayName: 'Test 2',
    nativeToken: exports.etherToken,
    publicRpcUrls: [{ http: 'http://localhost:8545' }],
    blockExplorers: [],
    blocks: {
        confirmations: 1,
        reorgPeriod: 1,
        estimateBlockTime: 3,
    },
    isTestnet: true,
};
exports.test3 = {
    chainId: 13373,
    name: chains_1.Chains.test3,
    protocol: ProtocolType.Ethereum,
    displayName: 'Test 3',
    nativeToken: exports.etherToken,
    publicRpcUrls: [{ http: 'http://localhost:8545' }],
    blockExplorers: [],
    blocks: {
        confirmations: 1,
        reorgPeriod: 2,
        estimateBlockTime: 3,
    },
    isTestnet: true,
};
/**
 * Collection maps
 *
 * NOTE: When adding chains here, consider also adding the
 * corresponding chain logo images in the /sdk/logos/* folders
 */
exports.chainMetadata = {
    alfajores: exports.alfajores,
    arbitrum: exports.arbitrum,
    arbitrumgoerli: exports.arbitrumgoerli,
    avalanche: exports.avalanche,
    bsc: exports.bsc,
    bsctestnet: exports.bsctestnet,
    celo: exports.celo,
    ethereum: exports.ethereum,
    fuji: exports.fuji,
    goerli: exports.goerli,
    sepolia: exports.sepolia,
    moonbasealpha: exports.moonbasealpha,
    moonbeam: exports.moonbeam,
    mumbai: exports.mumbai,
    optimism: exports.optimism,
    optimismgoerli: exports.optimismgoerli,
    polygon: exports.polygon,
    gnosis: exports.gnosis,
    test1: exports.test1,
    test2: exports.test2,
    test3: exports.test3,
};
// For convenient use in wagmi-based apps
exports.wagmiChainMetadata = (0, objects_1.objMap)(exports.chainMetadata, (_, metadata) => (0, wagmi_1.chainMetadataToWagmiChain)(metadata));
exports.chainIdToMetadata = Object.values(exports.chainMetadata).reduce((result, chain) => {
    result[chain.chainId] = chain;
    return result;
}, {});
exports.mainnetChainsMetadata = chains_1.Mainnets.map((chainName) => exports.chainMetadata[chainName]);
exports.testnetChainsMetadata = chains_1.Testnets.map((chainName) => exports.chainMetadata[chainName]);
function isValidChainMetadata(c) {
    return exports.ChainMetadataSchema.safeParse(c).success;
}
exports.isValidChainMetadata = isValidChainMetadata;
