// @hyperlane-xyz/sdk/dist/consts/chains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllChains = exports.TestChains = exports.Testnets = exports.Mainnets = exports.AllDeprecatedChains = exports.DeprecatedChains = exports.Chains = void 0;
/**
 * Enumeration of Hyperlane supported chains
 * Must be string type to be used with Object.keys
 */
var Chains;
(function (Chains) {
    Chains["alfajores"] = "alfajores";
    Chains["arbitrum"] = "arbitrum";
    Chains["arbitrumgoerli"] = "arbitrumgoerli";
    Chains["avalanche"] = "avalanche";
    Chains["bsc"] = "bsc";
    Chains["bsctestnet"] = "bsctestnet";
    Chains["celo"] = "celo";
    Chains["ethereum"] = "ethereum";
    Chains["fuji"] = "fuji";
    Chains["goerli"] = "goerli";
    Chains["sepolia"] = "sepolia";
    Chains["moonbasealpha"] = "moonbasealpha";
    Chains["moonbeam"] = "moonbeam";
    Chains["mumbai"] = "mumbai";
    Chains["optimism"] = "optimism";
    Chains["optimismgoerli"] = "optimismgoerli";
    Chains["polygon"] = "polygon";
    Chains["gnosis"] = "gnosis";
    Chains["test1"] = "test1";
    Chains["test2"] = "test2";
    Chains["test3"] = "test3";
})(Chains = exports.Chains || (exports.Chains = {}));
var DeprecatedChains;
(function (DeprecatedChains) {
    DeprecatedChains["arbitrumkovan"] = "arbitrumkovan";
    DeprecatedChains["arbitrumrinkeby"] = "arbitrumrinkeby";
    DeprecatedChains["kovan"] = "kovan";
    DeprecatedChains["rinkeby"] = "rinkeby";
    DeprecatedChains["optimismkovan"] = "optimismkovan";
    DeprecatedChains["optimismrinkeby"] = "optimismrinkeby";
})(DeprecatedChains = exports.DeprecatedChains || (exports.DeprecatedChains = {}));
exports.AllDeprecatedChains = Object.keys(DeprecatedChains);
exports.Mainnets = [
    Chains.arbitrum,
    Chains.avalanche,
    Chains.bsc,
    Chains.celo,
    Chains.ethereum,
    Chains.moonbeam,
    Chains.optimism,
    Chains.polygon,
    Chains.gnosis,
];
exports.Testnets = [
    Chains.alfajores,
    Chains.arbitrumgoerli,
    Chains.bsctestnet,
    Chains.fuji,
    Chains.goerli,
    Chains.moonbasealpha,
    Chains.mumbai,
    Chains.optimismgoerli,
    Chains.sepolia,
];
exports.TestChains = [
    Chains.test1,
    Chains.test2,
    Chains.test3,
];
exports.AllChains = [
    ...exports.Mainnets,
    ...exports.Testnets,
    ...exports.TestChains,
];
