// @hyperlane-xyz/utils/dist/src/logging.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.error = exports.warn = exports.log = exports.debug = exports.trace = void 0;
const utils_1 = require("./utils");
const ENV_LOG_LEVEL = ((_a = (0, utils_1.safelyAccessEnvVar)('LOG_LEVEL')) !== null && _a !== void 0 ? _a : 'debug').toLowerCase();
const LOG_TRACE = ENV_LOG_LEVEL == 'trace';
const LOG_DEBUG = LOG_TRACE || ENV_LOG_LEVEL == 'debug';
const LOG_INFO = LOG_DEBUG || ENV_LOG_LEVEL == 'info';
const LOG_WARN = LOG_INFO || ENV_LOG_LEVEL == 'warn';
const LOG_ERROR = LOG_WARN || ENV_LOG_LEVEL == 'error';
function trace(message, data) {
    if (LOG_TRACE)
        logWithFunction(console.trace, 'trace', message, data);
}
exports.trace = trace;
function debug(message, data) {
    if (LOG_DEBUG)
        logWithFunction(console.debug, 'debug', message, data);
}
exports.debug = debug;
function log(message, data) {
    if (LOG_INFO)
        logWithFunction(console.log, 'info', message, data);
}
exports.log = log;
function warn(message, data) {
    if (LOG_WARN)
        logWithFunction(console.warn, 'warn', message, data);
}
exports.warn = warn;
function error(message, data) {
    if (LOG_ERROR)
        logWithFunction(console.error, 'error', message, data);
}
exports.error = error;
function logWithFunction(logFn, level, message, data) {
    const fullLog = Object.assign(Object.assign({}, data), { level,
        message });
    logFn(JSON.stringify(fullLog));
}
