// @hyperlane-xyz/core/dist/factories/contracts/interfaces/hooks/IMessageHook__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IMessageHook__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint32",
                name: "_destination",
                type: "uint32",
            },
            {
                internalType: "bytes32",
                name: "_messageId",
                type: "bytes32",
            },
        ],
        name: "postDispatch",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "payable",
        type: "function",
    },
];
class IMessageHook__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IMessageHook__factory = IMessageHook__factory;
IMessageHook__factory.abi = _abi;
