// @hyperlane-xyz/core/dist/factories/contracts/libs/StaticMOfNAddressSetFactory__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticMOfNAddressSetFactory__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address[]",
                name: "_values",
                type: "address[]",
            },
            {
                internalType: "uint8",
                name: "_threshold",
                type: "uint8",
            },
        ],
        name: "deploy",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address[]",
                name: "_values",
                type: "address[]",
            },
            {
                internalType: "uint8",
                name: "_threshold",
                type: "uint8",
            },
        ],
        name: "getAddress",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class StaticMOfNAddressSetFactory__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.StaticMOfNAddressSetFactory__factory = StaticMOfNAddressSetFactory__factory;
StaticMOfNAddressSetFactory__factory.abi = _abi;
