// @hyperlane-xyz/sdk/dist/utils/wagmi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chainMetadataToWagmiChain = void 0;
const chainMetadata_1 = require("../consts/chainMetadata");
function chainMetadataToWagmiChain(metadata) {
    var _a;
    return {
        id: metadata.chainId,
        name: metadata.displayName || metadata.name,
        network: metadata.name,
        nativeCurrency: metadata.nativeToken || chainMetadata_1.etherToken,
        rpcUrls: {
            public: { http: [metadata.publicRpcUrls[0].http] },
            default: { http: [metadata.publicRpcUrls[0].http] },
        },
        blockExplorers: ((_a = metadata.blockExplorers) === null || _a === void 0 ? void 0 : _a.length)
            ? {
                default: {
                    name: metadata.blockExplorers[0].name,
                    url: metadata.blockExplorers[0].url,
                },
            }
            : undefined,
        testnet: !!metadata.isTestnet,
    };
}
exports.chainMetadataToWagmiChain = chainMetadataToWagmiChain;
