// @hyperlane-xyz/core/dist/factories/contracts/test/bad-recipient/BadRecipient2__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BadRecipient2__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint32",
                name: "",
                type: "uint32",
            },
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        name: "handle",
        outputs: [],
        stateMutability: "pure",
        type: "function",
    },
];
const _bytecode = "0x6080604052348015600f57600080fd5b5060a78061001e6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c8063c122ad8d14602d575b600080fd5b603c6038366004603e565b5050565b005b60008060408385031215605057600080fd5b823563ffffffff81168114606357600080fd5b94602093909301359350505056fea264697066735822122008000fd96bbb2b602f15e146d060e148b12390beb3743a68247b654fbd44b99764736f6c63430008110033";
const isSuperArgs = (xs) => xs.length > 1;
class BadRecipient2__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.BadRecipient2__factory = BadRecipient2__factory;
BadRecipient2__factory.bytecode = _bytecode;
BadRecipient2__factory.abi = _abi;
